src/lib/y2network/widgets/boot_protocol.rb

```diff
@@ -0,0 +1,364 @@
+              Yast::Hostname.MergeFQ(Yast::DNS.hostname, Yast::DNS.domain)
+            )
+          end
+        when :bootproto_dynamic
+          static_enabled(false)
+          dynamic_enabled(true)
+          none_enabled(false)
+        when :bootproto_none
+          static_enabled(false)
+          dynamic_enabled(false)
+          none_enabled(true)
+        else
+          raise "Unexpected value for boot protocol #{value.inspect}"
+        end
+
+        nil
+      end
+
+      def store
+        # FIXME: this value reset should be in backend in general not Yast::UI responsibility
+        @settings["IPADDR"] = @settings["NETMASK"] = @settings["PREFIXLEN"] = ""
+        case value
+        when :bootproto_none
+          bootproto = "none"
+          if ibft_available?
+            bootproto = Yast::UI.QueryWidget(Id(:bootproto_ibft), :Value) ? "ibft" : "none"
+          end
+          @settings["BOOTPROTO"] = bootproto
+        when :bootproto_static
+          @settings["BOOTPROTO"] = "static"
+          @settings["IPADDR"] = Yast::UI.QueryWidget(:bootproto_ipaddr, :Value)
+          mask = Yast::UI.QueryWidget(:bootproto_netmask, :Value)
+          if mask.start_with?("/")
+            @settings["PREFIXLEN"] = mask[1..-1]
+          else
+            param = Yast::Netmask.Check6(mask) ? "PREFIXLEN" : "NETMASK"
+            @settings[param] = @mask
+          end
+          @settings["HOSTNAME"] = Yast::UI.QueryWidget(:bootproto_hostname, :Value)
+        when :bootproto_dynamic
+          case Yast::UI.QueryWidget(:bootproto_dyn, :Value)
+          when :bootproto_dhcp
+            case Yast::UI.QueryWidget(:bootproto_dhcp_mode, :Value)
+            when :bootproto_dhcp_both
+              @settings["BOOTPROTO"] = "dhcp"
+            when :bootproto_dhcp_v4
+              @settings["BOOTPROTO"] = "dhcp4"
+            when :bootproto_dhcp_v6
+              @settings["BOOTPROTO"] = "dhcp6"
+            else
+              raise "Unexpected dhcp mode value #{Yast::UI.QueryWidget(:bootproto_dhcp_mode, :Value).inspect}"
+            end
+          when :bootproto_dhcp_auto
+            @settings["BOOTPROTO"] = "dhcp+autoip"
+          when :bootproto_auto
+            @settings["BOOTPROTO"] = "autoip"
+          else
+            raise "Unexpected dynamic mode value #{Yast::UI.QueryWidget(:bootproto_dyn, :Value).inspect}"
+          end
+        else
+          raise "Unexpected boot protocol value #{Yast::UI.QueryWidget(:bootproto, :Value).inspect}"
+        end
+      end
+
+      def validate
+        return true if value != :bootproto_static
+
+        ipa = Yast::UI.QueryWidget(:bootproto_ipaddr, :Value)
+        if !ipa.empty? && !Yast::IP.Check(ipa)
+          Yast::Popup.Error(_("No valid IP address."))
+          Yast::UI.SetFocus(:bootproto_ipaddr)
+          return false
+        end
+
+        mask = Yast::UI.QueryWidget(:bootproto_netmask, :Value)
+        if ipa != "" && mask != "" && !valid_netmask(ipa, mask)
```

Comment:
NP: these checks for empty values can be removed. Validating functions should catch that ... or should be fixed to catch that.

Reply:
that is a bit problematic, as it means that we cannot leave tab without specifying it.

There are two possible approaches:

1. do not allow to leave tab until everything is correct
2. builder that validates its content and write error popup when user try to finish creating such device

Not such which one is better as both have advantages and disadvantages.

Reply:
ah, now I see. `ipa` does not need to be checked. and mask is first time checked here and here is validating function and really catched.